Build failed with Go 1.25

```
❯ go test -v ./...
=== RUN   TestAll
# net
/opt/homebrew/Cellar/go/1.25.1/libexec/src/net/file.go:22:11: undefined: fileConn
/opt/homebrew/Cellar/go/1.25.1/libexec/src/net/file.go:34:12: undefined: fileListener
/opt/homebrew/Cellar/go/1.25.1/libexec/src/net/file.go:46:11: undefined: filePacketConn
    main_test.go:26: failed to run command: exit status 1
--- FAIL: TestAll (0.58s)
FAIL
FAIL    github.com/otelwasm/wasibuilder 0.869s
?       github.com/otelwasm/wasibuilder/rules   [no test files]
FAIL
```

I checked the suspicious change that happened on Go 1.25, and I've found the commit below, [which is a change related to Windows support](https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e).

https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e:src/net/file_unix.go;dlc=8ec555931dc65f1cd0e7cf28d771615d8064a8ad
https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e:src/net/file_posix.go;dlc=8ec555931dc65f1cd0e7cf28d771615d8064a8ad

The commit moved the methods appearing in the error message from `file_unix.go` to `file_posix.go`.

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -13,6 +13,7 @@ jobs:
     runs-on: ubuntu-latest
     
     strategy:
+      fail-fast: false
       matrix:
         go-version: 
           - 1.23.9

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -19,6 +19,8 @@ jobs:
           - 1.24.0
           - 1.24.1
           - 1.24.2
+          - 1.25.0
+          - 1.25.1
 
     steps:
       - name: Checkout code

--- a/internal/version/version.go
+++ b/internal/version/version.go
@@ -1,0 +1,62 @@
+package version
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+)
+
+var goVersionRegex = regexp.MustCompile(`go(\d+)\.(\d+)\.(\d+)`)
+
+type Version struct {
+	Major int
+	Minor int
+	Patch int
+}
+
+func (v Version) Compare(other Version) int {
+	if v.Major != other.Major {
+		return v.Major - other.Major
+	}
+	if v.Minor != other.Minor {
+		return v.Minor - other.Minor
+	}
+	return v.Patch - other.Patch
+}
+
+// IsAtLeast returns true if this version is >= the specified version
+func (v Version) IsAtLeast(major, minor, patch int) bool {
+	return v.Compare(Version{Major: major, Minor: minor, Patch: patch}) >= 0
+}
+
+// GetGoVersion gets the Go version from the GOVERSION environment variable
+// This is the most reliable method as it's automatically set by the Go runtime
+func GetGoVersion() (Version, error) {
+	goVersion := os.Getenv("GOVERSION")
+	if goVersion == "" {
+		return Version{}, fmt.Errorf("GOVERSION environment variable not set")
+	}
+
+	matches := goVersionRegex.FindStringSubmatch(goVersion)
+	if len(matches) != 4 {
+		return Version{}, fmt.Errorf("failed to parse GOVERSION: %s", goVersion)
+	}
+
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return Version{}, fmt.Errorf("failed to parse major version from GOVERSION: %s", goVersion)
+	}
+
+	minor, err := strconv.Atoi(matches[2])
+	if err != nil {
+		return Version{}, fmt.Errorf("failed to parse minor version from GOVERSION: %s", goVersion)
+	}
+
+	patch, err := strconv.Atoi(matches[3])
+	if err != nil {
+		return Version{}, fmt.Errorf("failed to parse patch version from GOVERSION: %s", goVersion)
+	}
+
+	return Version{Major: major, Minor: minor, Patch: patch}, nil
+}

--- a/internal/version/version.go
+++ b/internal/version/version.go
@@ -26,8 +26,8 @@ func (v Version) Compare(other Version) int {
 }
 
 // IsAtLeast returns true if this version is >= the specified version
-func (v Version) IsAtLeast(major, minor, patch int) bool {
-	return v.Compare(Version{Major: major, Minor: minor, Patch: patch}) >= 0
+func (v Version) IsAtLeast(other Version) bool {
+	return v.Compare(other) >= 0
 }
 
 // GetGoVersion gets the Go version from the GOVERSION environment variable

--- a/main.go
+++ b/main.go
@@ -185,20 +185,30 @@ func modifyArgsIfNeeded(toolName string, args []string) ([]string, error) {
 		}
 	}
 
+	// Detect Go version from the toolchain
+	var goVersion *rules.GoVersion
+	if version, err := rules.DetectGoVersionFromToolchain(toolName); err == nil {
+		goVersion = &version
+		slog.Debug("Detected Go version from toolchain", "version", version, "toolchain", toolName)
+	} else {
+		slog.Warn("Failed to detect Go version from toolchain, proceeding without version info", "error", err, "toolchain", toolName)
+	}
+
 	eCtx := rules.ExecContext{
 		Command:      toolName,
 		Args:         args,
 		Package:      packagePath,
 		PackageIndex: packageIndex,
+		GoVersion:    goVersion,
 	}
 
 	for _, rule := range allRules {
 		logger := slog.With("rule", rule.Name())
 
-		logger.Debug("Applying", "rule", rule.Name())
+		logger.Debug("Applying rule", "rule", rule.Name(), "package", packagePath, "command", toolName)
 
 		if err := rule.Apply(&eCtx); err != nil {
-			logger.Error("Error applying rule", "rule", rule.Name(), "error", err)
+			logger.Error("Error applying rule", "rule", rule.Name(), "package", packagePath, "error", err)
 			return nil, err
 		}
 	}

--- a/main.go
+++ b/main.go
@@ -8,6 +8,7 @@ import (
 	"slices"
 	"strings"
 
+	"github.com/otelwasm/wasibuilder/internal/version"
 	"github.com/otelwasm/wasibuilder/rules"
 )
 
@@ -186,10 +187,10 @@ func modifyArgsIfNeeded(toolName string, args []string) ([]string, error) {
 	}
 
 	// Get Go version from environment variable
-	var goVersion *rules.GoVersion
-	if version, err := rules.GetGoVersion(); err == nil {
-		goVersion = &version
-		slog.Debug("Got Go version", "version", version)
+	var goVersion *version.Version
+	if v, err := version.GetGoVersion(); err == nil {
+		goVersion = &v
+		slog.Debug("Got Go version", "version", v)
 	} else {
 		slog.Warn("Failed to get Go version, proceeding without version info", "error", err)
 	}

--- a/main.go
+++ b/main.go
@@ -185,13 +185,13 @@ func modifyArgsIfNeeded(toolName string, args []string) ([]string, error) {
 		}
 	}
 
-	// Detect Go version from the toolchain
+	// Get Go version from environment variable
 	var goVersion *rules.GoVersion
-	if version, err := rules.DetectGoVersionFromToolchain(toolName); err == nil {
+	if version, err := rules.GetGoVersion(); err == nil {
 		goVersion = &version
-		slog.Debug("Detected Go version from toolchain", "version", version, "toolchain", toolName)
+		slog.Debug("Got Go version", "version", version)
 	} else {
-		slog.Warn("Failed to detect Go version from toolchain, proceeding without version info", "error", err, "toolchain", toolName)
+		slog.Warn("Failed to get Go version, proceeding without version info", "error", err)
 	}
 
 	eCtx := rules.ExecContext{

--- a/rules/rules.go
+++ b/rules/rules.go
@@ -1,12 +1,21 @@
 package rules
 
-import "slices"
+import (
+	"os/exec"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+)
+
+var goVersionRegex = regexp.MustCompile(`go(\d+)\.(\d+)\.(\d+)`)
 
 type ExecContext struct {
 	Command      string
 	Args         []string
 	Package      string
 	PackageIndex int
+	GoVersion    *GoVersion
 }
 
 func (ctx *ExecContext) Clone() *ExecContext {
@@ -15,10 +24,131 @@ func (ctx *ExecContext) Clone() *ExecContext {
 		Args:         slices.Clone(ctx.Args),
 		Package:      ctx.Package,
 		PackageIndex: ctx.PackageIndex,
+		GoVersion:    ctx.GoVersion,
 	}
 }
 
 type Rule interface {
 	Apply(ctx *ExecContext) error
 	Name() string
+}
+
+// GoVersion represents a Go version with major, minor, and patch components
+type GoVersion struct {
+	Major int
+	Minor int
+	Patch int
+}
+
+// Compare returns -1 if v < other, 0 if v == other, 1 if v > other
+func (v GoVersion) Compare(other GoVersion) int {
+	if v.Major != other.Major {
+		if v.Major < other.Major {
+			return -1
+		}
+		return 1
+	}
+	if v.Minor != other.Minor {
+		if v.Minor < other.Minor {
+			return -1
+		}
+		return 1
+	}
+	if v.Patch != other.Patch {
+		if v.Patch < other.Patch {
+			return -1
+		}
+		return 1
+	}
+	return 0
+}
+
+// IsAtLeast returns true if this version is >= the specified version
+func (v GoVersion) IsAtLeast(major, minor, patch int) bool {
+	other := GoVersion{Major: major, Minor: minor, Patch: patch}
+	return v.Compare(other) >= 0
+}
+
+// DetectGoVersion detects the Go version being used by the toolchain
+func DetectGoVersion(goCommand string) (GoVersion, error) {
+	cmd := exec.Command(goCommand, "version")
+	output, err := cmd.Output()
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	// Parse output like "go version go1.25.1 darwin/arm64"
+	versionStr := strings.TrimSpace(string(output))
+
+	// Extract version using precompiled regex
+	matches := goVersionRegex.FindStringSubmatch(versionStr)
+
+	if len(matches) != 4 {
+		return GoVersion{}, err
+	}
+
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	minor, err := strconv.Atoi(matches[2])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	patch, err := strconv.Atoi(matches[3])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	return GoVersion{
+		Major: major,
+		Minor: minor,
+		Patch: patch,
+	}, nil
+}
+
+// DetectGoVersionFromToolchain detects the Go version directly from the toolchain binary
+// It uses the tool's own version flag (e.g., compile -V) to get version information
+func DetectGoVersionFromToolchain(toolPath string) (GoVersion, error) {
+	// Try to get version directly from the tool using -V flag
+	cmd := exec.Command(toolPath, "-V")
+	output, err := cmd.Output()
+	if err != nil {
+		// Fallback to detecting from "go" command in PATH
+		return DetectGoVersion("go")
+	}
+
+	// Parse output like "compile version go1.25.1"
+	versionStr := strings.TrimSpace(string(output))
+
+	// Extract version using precompiled regex
+	matches := goVersionRegex.FindStringSubmatch(versionStr)
+
+	if len(matches) != 4 {
+		// Fallback to detecting from "go" command in PATH
+		return DetectGoVersion("go")
+	}
+
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	minor, err := strconv.Atoi(matches[2])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	patch, err := strconv.Atoi(matches[3])
+	if err != nil {
+		return GoVersion{}, err
+	}
+
+	return GoVersion{
+		Major: major,
+		Minor: minor,
+		Patch: patch,
+	}, nil
 }

--- a/rules/rules.go
+++ b/rules/rules.go
@@ -2,11 +2,10 @@ package rules
 
 import (
 	"fmt"
-	"os/exec"
+	"os"
 	"regexp"
 	"slices"
 	"strconv"
-	"strings"
 )
 
 var goVersionRegex = regexp.MustCompile(`go(\d+)\.(\d+)\.(\d+)`)
@@ -70,81 +69,34 @@ func (v GoVersion) IsAtLeast(major, minor, patch int) bool {
 	return v.Compare(other) >= 0
 }
 
-// DetectGoVersion detects the Go version being used by the toolchain
-func DetectGoVersion(goCommand string) (GoVersion, error) {
-	cmd := exec.Command(goCommand, "version")
-	output, err := cmd.Output()
-	if err != nil {
-		return GoVersion{}, err
+// GetGoVersion gets the Go version from the GOVERSION environment variable
+// This is the most reliable method as it's automatically set by the Go runtime
+func GetGoVersion() (GoVersion, error) {
+	goVersion := os.Getenv("GOVERSION")
+	if goVersion == "" {
+		return GoVersion{}, fmt.Errorf("GOVERSION environment variable not set")
 	}
 
-	// Parse output like "go version go1.25.1 darwin/arm64"
-	versionStr := strings.TrimSpace(string(output))
-
 	// Extract version using precompiled regex
-	matches := goVersionRegex.FindStringSubmatch(versionStr)
+	matches := goVersionRegex.FindStringSubmatch(goVersion)
 
 	if len(matches) != 4 {
-		return GoVersion{}, fmt.Errorf("failed to extract version from version string")
+		return GoVersion{}, fmt.Errorf("failed to parse GOVERSION: %s", goVersion)
 	}
 
 	major, err := strconv.Atoi(matches[1])
 	if err != nil {
-		return GoVersion{}, err
+		return GoVersion{}, fmt.Errorf("invalid major version in GOVERSION: %s", goVersion)
 	}
 
 	minor, err := strconv.Atoi(matches[2])
 	if err != nil {
-		return GoVersion{}, err
+		return GoVersion{}, fmt.Errorf("invalid minor version in GOVERSION: %s", goVersion)
 	}
 
 	patch, err := strconv.Atoi(matches[3])
 	if err != nil {
-		return GoVersion{}, err
-	}
-
-	return GoVersion{
-		Major: major,
-		Minor: minor,
-		Patch: patch,
-	}, nil
-}
-
-// DetectGoVersionFromToolchain detects the Go version directly from the toolchain binary
-// It uses the tool's own version flag (e.g., compile -V) to get version information
-func DetectGoVersionFromToolchain(toolPath string) (GoVersion, error) {
-	// Try to get version directly from the tool using -V flag
-	cmd := exec.Command(toolPath, "-V")
-	output, err := cmd.Output()
-	if err != nil {
-		// Fallback to detecting from "go" command in PATH
-		return DetectGoVersion("go")
-	}
-
-	// Parse output like "compile version go1.25.1"
-	versionStr := strings.TrimSpace(string(output))
-
-	// Extract version using precompiled regex
-	matches := goVersionRegex.FindStringSubmatch(versionStr)
-
-	if len(matches) != 4 {
-		// Fallback to detecting from "go" command in PATH
-		return DetectGoVersion("go")
-	}
-
-	major, err := strconv.Atoi(matches[1])
-	if err != nil {
-		return GoVersion{}, err
-	}
-
-	minor, err := strconv.Atoi(matches[2])
-	if err != nil {
-		return GoVersion{}, err
-	}
-
-	patch, err := strconv.Atoi(matches[3])
-	if err != nil {
-		return GoVersion{}, err
+		return GoVersion{}, fmt.Errorf("invalid patch version in GOVERSION: %s", goVersion)
 	}
 
 	return GoVersion{

--- a/rules/rules.go
+++ b/rules/rules.go
@@ -1,21 +1,17 @@
 package rules
 
 import (
-	"fmt"
-	"os"
-	"regexp"
 	"slices"
-	"strconv"
-)
 
-var goVersionRegex = regexp.MustCompile(`go(\d+)\.(\d+)\.(\d+)`)
+	"github.com/otelwasm/wasibuilder/internal/version"
+)
 
 type ExecContext struct {
 	Command      string
 	Args         []string
 	Package      string
 	PackageIndex int
-	GoVersion    *GoVersion
+	GoVersion    *version.Version
 }
 
 func (ctx *ExecContext) Clone() *ExecContext {
@@ -31,77 +27,4 @@ func (ctx *ExecContext) Clone() *ExecContext {
 type Rule interface {
 	Apply(ctx *ExecContext) error
 	Name() string
-}
-
-// GoVersion represents a Go version with major, minor, and patch components
-type GoVersion struct {
-	Major int
-	Minor int
-	Patch int
-}
-
-// Compare returns -1 if v < other, 0 if v == other, 1 if v > other
-func (v GoVersion) Compare(other GoVersion) int {
-	if v.Major != other.Major {
-		if v.Major < other.Major {
-			return -1
-		}
-		return 1
-	}
-	if v.Minor != other.Minor {
-		if v.Minor < other.Minor {
-			return -1
-		}
-		return 1
-	}
-	if v.Patch != other.Patch {
-		if v.Patch < other.Patch {
-			return -1
-		}
-		return 1
-	}
-	return 0
-}
-
-// IsAtLeast returns true if this version is >= the specified version
-func (v GoVersion) IsAtLeast(major, minor, patch int) bool {
-	other := GoVersion{Major: major, Minor: minor, Patch: patch}
-	return v.Compare(other) >= 0
-}
-
-// GetGoVersion gets the Go version from the GOVERSION environment variable
-// This is the most reliable method as it's automatically set by the Go runtime
-func GetGoVersion() (GoVersion, error) {
-	goVersion := os.Getenv("GOVERSION")
-	if goVersion == "" {
-		return GoVersion{}, fmt.Errorf("GOVERSION environment variable not set")
-	}
-
-	// Extract version using precompiled regex
-	matches := goVersionRegex.FindStringSubmatch(goVersion)
-
-	if len(matches) != 4 {
-		return GoVersion{}, fmt.Errorf("failed to parse GOVERSION: %s", goVersion)
-	}
-
-	major, err := strconv.Atoi(matches[1])
-	if err != nil {
-		return GoVersion{}, fmt.Errorf("invalid major version in GOVERSION: %s", goVersion)
-	}
-
-	minor, err := strconv.Atoi(matches[2])
-	if err != nil {
-		return GoVersion{}, fmt.Errorf("invalid minor version in GOVERSION: %s", goVersion)
-	}
-
-	patch, err := strconv.Atoi(matches[3])
-	if err != nil {
-		return GoVersion{}, fmt.Errorf("invalid patch version in GOVERSION: %s", goVersion)
-	}
-
-	return GoVersion{
-		Major: major,
-		Minor: minor,
-		Patch: patch,
-	}, nil
 }

--- a/rules/rules.go
+++ b/rules/rules.go
@@ -1,6 +1,7 @@
 package rules
 
 import (
+	"fmt"
 	"os/exec"
 	"regexp"
 	"slices"
@@ -84,7 +85,7 @@ func DetectGoVersion(goCommand string) (GoVersion, error) {
 	matches := goVersionRegex.FindStringSubmatch(versionStr)
 
 	if len(matches) != 4 {
-		return GoVersion{}, err
+		return GoVersion{}, fmt.Errorf("failed to extract version from version string")
 	}
 
 	major, err := strconv.Atoi(matches[1])

--- a/rules/wasmedgenet.go
+++ b/rules/wasmedgenet.go
@@ -6,6 +6,8 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/otelwasm/wasibuilder/internal/version"
 )
 
 //go:embed wasmedgenet/*
@@ -86,7 +88,7 @@ func (w *WASMEdgeNet) processNetPackage(ctx *ExecContext, logger *slog.Logger) e
 	// Add file_posix.go only for Go 1.25 and above
 	// Required for Go 1.25 because of upstream refactoring.
 	// Ref, https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e
-	if ctx.GoVersion != nil && ctx.GoVersion.IsAtLeast(1, 25, 0) {
+	if ctx.GoVersion != nil && ctx.GoVersion.IsAtLeast(version.Version{Major: 1, Minor: 25, Patch: 0}) {
 		addedFiles = append(addedFiles, "src/net/file_posix.go")
 		logger.Debug("Adding file_posix.go for Go 1.25+", "package", ctx.Package, "version", ctx.GoVersion)
 	} else {

--- a/rules/wasmedgenet.go
+++ b/rules/wasmedgenet.go
@@ -81,11 +81,18 @@ func (w *WASMEdgeNet) processNetPackage(ctx *ExecContext, logger *slog.Logger) e
 		"src/net/fd_unix.go",
 		"src/net/sock_posix.go",
 		"src/net/file_unix.go",
-
-		// Required for Go 1.25 because of upstream refactoring.
-		// Ref, https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e
-		"src/net/file_posix.go",
 	}
+
+	// Add file_posix.go only for Go 1.25 and above
+	// Required for Go 1.25 because of upstream refactoring.
+	// Ref, https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e
+	if ctx.GoVersion != nil && ctx.GoVersion.IsAtLeast(1, 25, 0) {
+		addedFiles = append(addedFiles, "src/net/file_posix.go")
+		logger.Debug("Adding file_posix.go for Go 1.25+", "package", ctx.Package, "version", ctx.GoVersion)
+	} else {
+		logger.Debug("Skipping file_posix.go for Go < 1.25", "package", ctx.Package, "version", ctx.GoVersion)
+	}
+
 	var addedFilesFromFS = []string{
 		"net/fake.go.added",
 		"net/sockopt_wasip1.go.added",

--- a/rules/wasmedgenet.go
+++ b/rules/wasmedgenet.go
@@ -81,6 +81,10 @@ func (w *WASMEdgeNet) processNetPackage(ctx *ExecContext, logger *slog.Logger) e
 		"src/net/fd_unix.go",
 		"src/net/sock_posix.go",
 		"src/net/file_unix.go",
+
+		// Required for Go 1.25 because of upstream refactoring.
+		// Ref, https://cs.opensource.google/go/go/+/6953ef86cd72a835d398319c4da560c8b78ba28e
+		"src/net/file_posix.go",
 	}
 	var addedFilesFromFS = []string{
 		"net/fake.go.added",

